### ISSUE TITLE
Property lookups on arguments to type parameters constrained by string index signatures can violate other constraints because undefined is included for optional properties

### ISSUE THREAD

**Issue by aweebit** [microsoft/TypeScript#63709]
### 🔎 Search Terms

index signature, indexed types, mapped types, lookup types, optional properties, generic constraints

### 🕗 Version & Regression Information

- It's only been possible to assign optional properties to string index signature since version 4.2 (see #41921)
- The bug has existed ever since

### ⏯ Playground Link

https://www.typescriptlang.org/play/?#code/C4TwDgpgBAkgznArhAPAFShAHsCA7AEzigG8oBtLALijmACcBLPAcwF0a9EBbAIwnoBuKAF8AfFAC8UAHI8BjAMbpyAawggA9gDMoaNmMEBYAFChIs+U2UzMOfEShc+AidJnGz4aAA0psBGQUMgBDAH5OHn56UUMgA

### 💻 Code

```ts
type Issue<T extends { [x: string]: number; }> = Numeric<T[keyof T]>;
type Numeric<N extends number> = N;
type X = Issue<{ a?: 0 }>;
```

### 🙁 Actual behavior

During the evaluation of `X`, `T[keyof T]` becomes `{ a?: 0 }['a']`, which in turn is evaluated as `0 | undefined` because `a` is an optional property. Therefore, the argument passed to `Numeric` is `0 | undefined`. This type doesn't satisfy the `extends number` constraint and therefore should not be permitted. However, the compiler doesn't report any errors.

### 🙂 Expected behavior

An error is reported on line 1, saying that the `N extends number` constraint might be violated.

### Additional information about the issue

The same happens with the mapped type `{ [x in string]: number; }`.

The problem only exists for string index signatures: for both `{ [x: number]: number; }` and `{ [x: symbol]: number; }`, there are errors reported.

#41921 might've been a mistake. Perhaps a new tsconfig option could be introduced that would revert the change so that the rule for assignment of optional properties to string index signatures is more strict and consistent to how number / symbol index signatures are handled again? With `"exactOptionalPropertyTypes": true` at least, the original issue #41418 should not cause trouble anymore. Actually, I think that the introduction of `exactOptionalPropertyTypes` is exactly how the issue should have been solved back then. Unfortunately, the setting was only introduced later (in version 4.4).

**Comment by RyanCavanaugh** [comment]
I don't see a fix here that wouldn't make a lot of very "this code is fine" examples start complaining. If you have one I'm all ears, but this particular cure seems worse than the current disease.

**Comment by aweebit** [comment]
I've thought a little more about the issue.

First of all, I don't think #41921 was a mistake after all. Here is my reasoning:
- Interpreting `{ [key: string]: boolean }` as the type for objects that always give you a boolean when a property with a string key is accessed would not be particularly useful since such objects can only be created with `new Proxy(...)` and are extremely rare. However, this is pretty much how the type is interpreted without `noUncheckedIndexedAccess`
- Instead, the type should be understood as a constraint for more precise types whose all string keys are mapped to boolean values
- If instead of `string` we had some bounded key type `Key` such as `'a' | 'b' | 'c'`, this is how such a constraint would look like: `{ [K in Key]?: boolean }`
- We have to explicitly mark the mapped keys as optional so that types like `{ a: boolean; b?: boolean }` where all `Key` keys are still mapped to boolean values despite some of them being optional don't fail the constraint
- With index signatures, optionality should be assumed without it being specified explicitly
- This is exactly the treatment [#41921](https://github.com/microsoft/TypeScript/pull/41921) established for string index signatures
- For reasons of soundness and consistency, I think that the same assumption should be made for number and symbol index signatures too, i.e. this example from [#41921](https://github.com/microsoft/TypeScript/pull/41921) should not fail:

```ts
declare let sortOfArrayish: { [key: number]: string };
declare let numberKeys: { 42?: string };
sortOfArrayish = numberKeys;
```

This brings us back to the issue I've reported here. If #41921 wasn't a mistake, how should the issue be solved without reverting that change?

My answer is that a new tsconfig option should be introduced that would enable strict checking of lookup types. It could be called `strictLookupTypes`. Here is what it would do:
- Disallow constraint violations due to optionality reasons like the one in my example
- Cause the compiler to no longer assume that a type constrained by `{ [x: string]: Something; }` will only have string keys when a lookup on it is checked against a constraint like it apparently does now because otherwise, it would also report an error on line 1 of my example because `T[keyof T]` could violate the `extends number` constraint like it does when `T` is `{ [sym]: boolean }` (where `sym` is some symbol variable), and so `T[keyof T]` is `boolean` and therefore should not be accepted by `Numeric`, but there is currently no error reported for this, too (I've only discovered this now, but the fact that `Issue<{ a?: 0; [sym]: boolean }>` evaluates to `boolean | 0 | undefined` with no errors is absolutely wild)
- Disallow other constraint violations by lookup types, ideally including those described by me in #62665 and by @mkantor in #55733

@RyanCavanaugh Should I create a new feature request issue for this?

**Comment by aweebit** [comment]
Pretty much the same problem was demonstrated in https://github.com/microsoft/TypeScript/issues/48069#issuecomment-1201974238:

> ```ts
> type Original = {p1?: string}
> 
> // seems equivalent to keyof but isn't under strictNullChecks because `undefined` will be included if T has any optional properties
> type CustomKeyOf<T> = { [K in keyof T]: K }[keyof T]
> 
> type CorrectlyErrors = Pick<Original, CustomKeyOf<Original>> // Type '"p1" | undefined' does not satisfy the constraint '"p1"'
> 
> // This should error like CorrectlyErrors, but making it generic currently allows it to slip by
> // I believe this is ultimately the root problem because it results in `undefined` being considered a key below
> type Identity<T> = Pick<T, CustomKeyOf<T>>
> ```

It's the same because just like here, TypeScript ignores the fact that `undefined` could be included in a lookup type due to optional keys when checking it against a constraint.

In my example, the lookup type is `T[keyof T]`, while the constraint is `N extends number` and comes from the `Numeric` type.

In this other example, the lookup type is `{ [K in keyof T]: K }[keyof T]`, while the constraint is `K extends keyof T` and comes from the built-in `Pick` type.

In the linked issue, the fact that no errors are reported in such cases leads to a very confusing problem that is extremely difficult to debug.
